pkg/controlplane/peer_endpoint_controller.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controlplane
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	utilpeerproxy "k8s.io/apiserver/pkg/util/peerproxy"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controlplane/reconcilers"
+	corerest "k8s.io/kubernetes/pkg/registry/core/rest"
+	"k8s.io/kubernetes/pkg/util/async"
+)
+
+// PeerEndpointController is the controller manager for updating the Kubernetes service
+// endpoints. This is essentially a clone of the controller manager for the core
+// bootstrap Kubernetes controller loops.
+
+// We have cloned the controller to have a separate independent endpoint reconciliation loop
+// which ensures that the peer kube-apiservers are fetching the updated endpoint info for a given apiserver
+// in the case when the peer wants to proxy the request to the given apiserver because it can not serve the
+// request itself due to version mismatch.
+type PeerEndpointController struct {
```

Comment:
We decided to use only the lease controller NewLeaseEndpointReconciler  pkg/controlplane/reconcilers/lease.go , not the endpoint controller, https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/controlplane/reconcilers/lease.go 

https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4020-unknown-version-interoperability-proxy#identifying-destination-apiservers-network-location

we don't want to publish Endpoint and EndpointSlices :)

Reply:
yeah I had doubts about this part of the code. Thanks for a thorough review!
I have made changes and removed the publishing endpoint/endpointslices bit. Could you take a second look?

Reply:
I made a new PeerEndpointLeaseController that uses the PeerEndpointReconciler to update peer leases.